### PR TITLE
Rfink/java8 optional

### PR DESCRIPTION
**Pull request by uschi2000**
This change is [Reviewable](https://reviewable.io/reviews/palantir/http-remoting/321)

### CODE REVIEW

File: extras/jackson-support/build.gradle [hunk 1]
```diff
@@ -5,6 +5,7 @@ dependencies {
     compile "com.fasterxml.jackson.datatype:jackson-datatype-guava"
     compile "com.fasterxml.jackson.datatype:jackson-datatype-jdk7"
     compile "com.fasterxml.jackson.module:jackson-module-afterburner"
+    compile "com.fasterxml.jackson.datatype:jackson-datatype-jdk8"
```
Comment (markelliot) [inline extras/jackson-support/build.gradle:8]: also worthwhile to add `com.fasterxml.jackson.datatype:jackson-datatype-jsr310`, since that fixes #319

**Comment by uschi2000** [comment]
yep, will do. this pr needs a bit more work anyway

File: extras/jackson-support/build.gradle (hunk 1 above)
Comment (markelliot) [reply extras/jackson-support/build.gradle:8]: corresponding module class name is `JavaTimeModule`

**Comment by uschi2000** [comment]
@markelliot PR is pretty unwieldy, not quite sure how to reduce size :(

File: jaxrs-clients/src/main/java/feign/GoogleOptionalAwareDecoder.java [hunk 2]
```diff
@@ -28,11 +28,11 @@
  * Decorates a Feign {@link Decoder} such that it returns {@link Optional#absent} when observing an HTTP 204 error code
  * for a method with {@link Type} {@link Optional}.
  */
-public final class OptionalAwareDecoder implements Decoder {
+public final class GoogleOptionalAwareDecoder implements Decoder {
```
Comment (markelliot) [inline jaxrs-clients/src/main/java/feign/GoogleOptionalAwareDecoder.java]: `GuavaOptionalAwareDecoder` for consistency with rest of code

**Comment by markelliot** [comment]
I think you're missing the retrofit equivalents of the feign stuff?

**Comment by uschi2000** [comment]
Will do in separate PR. https://github.com/palantir/http-remoting/issues/326

File: jaxrs-clients/src/test/java/com/palantir/remoting1/jaxrs/feignimpl/Java8TestServer.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        @Override
+        public Java8ComplexType getJava8ComplexType(Java8ComplexType value) {
+            return value;
+        }
+    }
+
+    @Path("/")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public interface TestService {
```
Comment (markelliot) [inline jaxrs-clients/src/test/java/com/palantir/remoting1/jaxrs/feignimpl/Java8TestServer.java:150]: I think it'd be wise to add some test cases for accepting a `QueryParam` that's `Optional`

**Comment by markelliot** [comment]
ok, mostly fine with this, missing tests for params that are `Optional`

**Comment by uschi2000** [comment]
Added tests for Optional parameters

**Pull request merged**